### PR TITLE
Partitioned heat equation scenario using waveform relaxation

### PR DESCRIPTION
**Pull request by BenjaminRueth**
This PR is strongly related to the features introduced in https://github.com/precice/fenics-adapter/pull/35.

* `heat.py` is modified such that waveform relaxation is used.
* Adapter and preCICE configuration files are modified correspondingly.

### CODE REVIEW

**Comment by BenjaminRueth** [comment]
Possibly, we can remove all the waveform relaxation specific config files (`precice-config-WRXX.xml`, `precice-adapter-config-D/N-WRXX`) and provide instructions on how to generate them using the script in `tools`.

File: HT/partitioned-heat/fenics-fenics/heat.py [hunk 1]
```diff
@@ -117,31 +106,26 @@ def fluxes_from_temperature_full_domain(F, V):
...
+d_subcycling = "D".format(wr_tag=wr_tag)
+n_subcycling = "N".format(wr_tag=wr_tag)
+
+configs_path = os.path.join("experiments", wr_tag, window_size)
 
 if problem is ProblemType.DIRICHLET:
     nx = nx*3
-    adapter_config_filename = "precice-adapter-config-D.json"
+    adapter_config_filename = os.path.join(configs_path, "precice-adapter-config-D.json")
+    other_adapter_config_filename = os.path.join(configs_path, "precice-adapter-config-N.json")
```
Comment (uekerman) [inline HT/partitioned-heat/fenics-fenics/heat.py]: Why is this needed?

File: HT/partitioned-heat/fenics-fenics/README.md [hunk 2]
```diff
@@ -1,3 +1,36 @@
 # Tutorial for a partitioned heat equation using FEniCS
 
 This tutorial is described in the [preCICE wiki](https://github.com/precice/precice/wiki/Tutorial-for-solving-the-heat-equation-in-a-partitioned-fashion-using-FEniCS).
+
+## Waveform relaxation (experimental)
+
+This branch of the tutorial uses the fenics-adapter waveform bindings. If you want to run this tutorial, you have to install the version of the FEniCS-adapter that can be found on [this branch](https://github.com/precice/fenics-adapter/tree/WaveformBindingsDraft). Further information on waveform relaxation in the FEniCS-adapter can be found in issue https://github.com/precice/fenics-adapter/issues/16 and milestone https://github.com/precice/fenics-adapter/milestone/2.
```
[the remark this answers is not in the document]
Comment (BenjaminRueth) [reply HT/partitioned-heat/fenics-fenics/README.md]: Thanks for the hint!

File: HT/partitioned-heat/fenics-fenics/README.md [hunk 3]
```diff
@@ -1,3 +1,36 @@
...
+For all setups linear interpolation in time is used for each (micro-)timestep. The commands for running the example cases are given.
+
+**WR11:** this case uses no subcycling such that `dt_neumann = dt_dirichlet = T_window` (dt for each solver equals window size).
+
+* ` python3 heat.py -d`
+* ` python3 heat.py -n`
+
+**WR12:** this case uses subcycling such that `2*dt_neumann = dt_dirichlet = T_window`.
+
+* ` python3 heat.py -d -wr 1 2`
```
[the remark this answers is not in the document]
Comment (BenjaminRueth) [reply HT/partitioned-heat/fenics-fenics/README.md]: The participants exchange the substeps via different pieces of data (`Temperature1`, `Temperature2` and `Flux1` in the case `-wr 1 2`). Both participants have to know the complete setup. If a participant only knew its own number of substeps, it could not know the data names that the other participant uses and therefore it could not receive the subsamples.

File: HT/partitioned-heat/fenics-fenics/experiments/WR11/dT0.1/precice-adapter-config-D.json [hunk 4]
```diff
@@ -5,5 +5,8 @@
       "coupling_mesh_name": "DirichletNodes",
       "write_data_name": "Flux",
       "read_data_name": "Temperature"
+    },
+  "waveform": {
+      "n_substeps": 1
```
[the remark this answers is not in the document]
Comment (BenjaminRueth) [reply HT/partitioned-heat/fenics-fenics/experiments/WR11/dT0.1/precice-adapter-config-D.json]: This is indeed redundant. What I said [above](https://github.com/precice/tutorials/pull/30/files/344bf6b3511c2ced79103120b864c7f1ce41670f#r294784543) is actually true for the config file. This means that the adapter has to know both config files in order to know the complete WR setup.

File: HT/partitioned-heat/fenics-fenics/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (BenjaminRueth) [reply HT/partitioned-heat/fenics-fenics/README.md]: See (here)[https://github.com/precice/tutorials/pull/30/files/344bf6b3511c2ced79103120b864c7f1ce41670f#r294792652]: The command line input is only used for creating the path of the configuration files.

File: HT/partitioned-heat/fenics-fenics/heat.py (hunk 1 above)
Comment (BenjaminRueth) [reply HT/partitioned-heat/fenics-fenics/heat.py]: This is needed for my setup where I want to run multiple cases, one after another. The configuration files are stored in a folder hierarchy.

File: HT/partitioned-heat/fenics-fenics/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (uekerman) [reply HT/partitioned-heat/fenics-fenics/README.md]: Got it. Thanks. Once WR is in preCICE, it should thus not be necessary.

File: HT/partitioned-heat/fenics-fenics/heat.py (hunk 1 above)
Comment (uekerman) [reply HT/partitioned-heat/fenics-fenics/heat.py]: I meant the "other", but you explained it above.
